### PR TITLE
Add reading and writing to LDS via threadwise_*

### PR DESCRIPTION
**Pull request by manupak**
This commit adds writing to LDS using the
threadwise_read_into and threadwise_write_all ops.

closes : https://github.com/ROCmSoftwarePlatform/llvm-project-private/issues/894

### CODE REVIEW

File: mlir/lib/Dialect/Rock/IR/RockDialect.cpp [hunk 1]
```diff
@@ -1435,16 +1435,35 @@ LogicalResult ThreadwiseReadIntoOp::verify() {
...
     inputShape = extraViews[0].cast<TransformMapAttr>().getUpperBounds();
 
-  if (inputShape.size() != 3)
-    return emitOpError("source view must accept (bid, tid, iter) coordinates");
+  MemRefType srcType = getSource().getType().cast<MemRefType>();
+  Attribute srcMemSpaceAttr = srcType.getMemorySpace();
+  // Unless specified it is assumed to be global
+  gpu::AddressSpace srcGpuMemSpace = gpu::AddressSpace::Global;
+  if (srcMemSpaceAttr) {
+    srcGpuMemSpace = srcMemSpaceAttr.cast<gpu::AddressSpaceAttr>().getValue();
```
Comment (krzysz00) [inline mlir/lib/Dialect/Rock/IR/RockDialect.cpp]: `auto srcMemSpaceAttr = srcType.getMemorySpace.dyn_cast_or_null<gpu::AddressSpaceAttr>();`
and then the if statement. Or possibly erroring out on memory spaces that aren't either absent or the GPU ones.
That cast<> will cause crashes on user input as opposed to programmer error.

File: mlir/lib/Dialect/Rock/Transforms/BlockwiseGemmToThreadwise.cpp [hunk 2]
```diff
@@ -738,21 +793,46 @@ LogicalResult ThreadwiseWriteAllRewritePattern::matchAndRewrite(
...
+      b.create<GlobalStoreOp>(
+          loc, source, buffer, b.getIndexAttr(vectorLen), op.getFeaturesAttr(),
+          op.getStoreMethodAttr(),
+          outLoop.getLowerCoords(/*domain=*/0)[iterDim[dstAddrSpace]],
+          outLoop.getValidity(/*domain=*/1),
+          outLoop.getLowerCoords(/*domain=*/1));
+    } else {
+      Type loadType = vectorTypeOrSelf(elementType, vectorLen);
+      TypedValue<IntegerType> valid = outLoop.getValidity(/*domain=*/0);
+      Value zeroInt = createZeroConstantOp(b, loc, valid.getType());
```
Comment (krzysz00) [inline mlir/lib/Dialect/Rock/Transforms/BlockwiseGemmToThreadwise.cpp]: This `cmp` is reduntant, - `validity` is already bool-typed

File: mlir/lib/Dialect/Rock/Transforms/BlockwiseGemmToThreadwise.cpp [hunk 3]
```diff
@@ -686,21 +699,52 @@ LogicalResult ThreadwiseReadIntoRewritePattern::matchAndRewrite(
...
+      Value isValid = b.createOrFold<arith::CmpIOp>(
+          loc, arith::CmpIPredicate::ne, valid, zeroInt);
+      scf::IfOp ifb =
+          b.create<scf::IfOp>(loc, isValid, /*withElseRegion=*/true);
+      {
+        OpBuilder thenb = ifb.getThenBodyBuilder(b.getListener());
+        Value loaded = thenb.create<InBoundsLoadOp>(
+            loc, loadType, buffer, loadLoop.getLowerCoords(/*domain=*/0));
+        thenb.create<InBoundsStoreOp>(
```
Comment (krzysz00) [inline mlir/lib/Dialect/Rock/Transforms/BlockwiseGemmToThreadwise.cpp]: Why not instead have the load be in the if, creating an if/else that yield's the value, and then unconditionally create the InBoundsStore?

**Review by krzysz00 — COMMENTED** [review]
Style nits more than fundamental objections

File: mlir/lib/Dialect/Rock/IR/RockDialect.cpp (hunk 1 above)
Comment (manupak) [reply mlir/lib/Dialect/Rock/IR/RockDialect.cpp]: good call! done.

File: mlir/lib/Dialect/Rock/Transforms/BlockwiseGemmToThreadwise.cpp (hunk 2 above)
Comment (manupak) [reply mlir/lib/Dialect/Rock/Transforms/BlockwiseGemmToThreadwise.cpp]: right! fixed.

File: mlir/lib/Dialect/Rock/Transforms/BlockwiseGemmToThreadwise.cpp (hunk 3 above)
Comment (manupak) [reply mlir/lib/Dialect/Rock/Transforms/BlockwiseGemmToThreadwise.cpp]: you are right. fixed now!

**Comment by manupak** [comment]
Thanks for the review @krzysz00! I believe they are addressed now.

File: mlir/lib/Dialect/Rock/Transforms/BlockwiseGemmToThreadwise.cpp [hunk 4]
```diff
@@ -664,13 +665,25 @@ LogicalResult ThreadwiseReadIntoRewritePattern::matchAndRewrite(
...
+  MemRefType srcBufferType = buffer.getType().cast<MemRefType>();
+  ArrayRef<int64_t> bufferShape = srcBufferType.getShape();
+
+  // Unless specified it is assumed to be global
+  gpu::AddressSpace srcAddrSpace = gpu::AddressSpace::Global;
+  if (srcBufferType.getMemorySpace()) {
+    srcAddrSpace =
+        srcBufferType.getMemorySpace().cast<gpu::AddressSpaceAttr>().getValue();
+  }
+  DenseMap<gpu::AddressSpace, size_t> iterDim{
```
Comment (krzysz00) [inline mlir/lib/Dialect/Rock/Transforms/BlockwiseGemmToThreadwise.cpp:677]: In silly minor notes, I think there's a `SmallDenseMap` we could use here? And I wonder if we can `constexpr` this

**Review by krzysz00 — APPROVED** [review]
Looks good to me!

**Pull request merged**